File: src/brs/web/api/http/handler/GetBlock.java
package brs.web.api.http.handler;

import brs.Block;
import brs.Blockchain;
import brs.web.api.http.common.LegacyDocTag;
import brs.web.api.http.common.Parameters;
import brs.services.BlockService;
import brs.util.Convert;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.JSONData;
import com.google.gson.JsonElement;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;

public final class GetBlock extends ApiServlet.JsonRequestHandler {

  private final Blockchain blockchain;

  private final BlockService blockService;

  public GetBlock(Blockchain blockchain, BlockService blockService) {
    super(new LegacyDocTag[] {LegacyDocTag.BLOCKS}, BLOCK_PARAMETER, HEIGHT_PARAMETER, TIMESTAMP_PARAMETER, INCLUDE_TRANSACTIONS_PARAMETER);
    this.blockchain = blockchain;
    this.blockService = blockService;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) {
    String blockValue = Convert.emptyToNull(req.getParameter(BLOCK_PARAMETER));
    String heightValue = Convert.emptyToNull(req.getParameter(HEIGHT_PARAMETER));
    String timestampValue = Convert.emptyToNull(req.getParameter(TIMESTAMP_PARAMETER));

    Block blockData;
    if (blockValue != null) {
      try {
        blockData = blockchain.getBlock(Convert.parseUnsignedLong(blockValue));
      } catch (RuntimeException e) {
        return INCORRECT_BLOCK;
      }
    } else if (heightValue != null) {
      try {
        int height = Integer.parseInt(heightValue);
        if (height < 0 || height > blockchain.getHeight()) {
          return INCORRECT_HEIGHT;
        }
        blockData = blockchain.getBlockAtHeight(height);
      } catch (RuntimeException e) {
        return INCORRECT_HEIGHT;
      }
    } else if (timestampValue != null) {
      try {
        int timestamp = Integer.parseInt(timestampValue);
        if (timestamp < 0) {
          return INCORRECT_TIMESTAMP;
        }
        blockData = blockchain.getLastBlock(timestamp);
      } catch (RuntimeException e) {
        return INCORRECT_TIMESTAMP;
      }
    } else {
      blockData = blockchain.getLastBlock();
    }

    if (blockData == null) {
      return UNKNOWN_BLOCK;
    }

    boolean includeTransactions = Parameters.isTrue(req.getParameter(INCLUDE_TRANSACTIONS_PARAMETER));

    return JSONData.block(blockData, includeTransactions, blockchain.getHeight(), blockService.getBlockReward(blockData), blockService.getScoopNum(blockData));
  }
}

File: src/main/kotlin/brs/http/GetBlock.kt
package brs.http

import brs.Blockchain
import brs.http.JSONResponses.INCORRECT_BLOCK
import brs.http.JSONResponses.INCORRECT_HEIGHT
import brs.http.JSONResponses.INCORRECT_TIMESTAMP
import brs.http.JSONResponses.UNKNOWN_BLOCK
import brs.http.common.Parameters.BLOCK_PARAMETER
import brs.http.common.Parameters.HEIGHT_PARAMETER
import brs.http.common.Parameters.INCLUDE_TRANSACTIONS_PARAMETER
import brs.http.common.Parameters.TIMESTAMP_PARAMETER
import brs.http.common.Parameters.isTrue
import brs.services.BlockService
import brs.util.convert.emptyToNull
import brs.util.convert.parseUnsignedLong
import com.google.gson.JsonElement

import javax.servlet.http.HttpServletRequest

internal class GetBlock internal constructor(private val blockchain: Blockchain, private val blockService: BlockService) : APIServlet.JsonRequestHandler(arrayOf(APITag.BLOCKS), BLOCK_PARAMETER, HEIGHT_PARAMETER, TIMESTAMP_PARAMETER, INCLUDE_TRANSACTIONS_PARAMETER) {

    override fun processRequest(request: HttpServletRequest): JsonElement {
        val blockValue = request.getParameter(BLOCK_PARAMETER).emptyToNull()
        val heightValue = request.getParameter(HEIGHT_PARAMETER).emptyToNull()
        val timestampValue = request.getParameter(TIMESTAMP_PARAMETER).emptyToNull()

        val blockData = when {
            blockValue != null -> try {
                blockchain.getBlock(blockValue.parseUnsignedLong())
            } catch (e: RuntimeException) {
                return INCORRECT_BLOCK
            }
            heightValue != null -> try {
                val height = Integer.parseInt(heightValue)
                if (height < 0 || height > blockchain.height) {
                    return INCORRECT_HEIGHT
                }
                blockchain.getBlockAtHeight(height)
            } catch (e: RuntimeException) {
                return INCORRECT_HEIGHT
            }
            timestampValue != null -> try {
                val timestamp = Integer.parseInt(timestampValue)
                if (timestamp < 0) {
                    return INCORRECT_TIMESTAMP
                }
                blockchain.getLastBlock(timestamp)
            } catch (e: RuntimeException) {
                return INCORRECT_TIMESTAMP
            }
            else -> blockchain.lastBlock
        } ?: return UNKNOWN_BLOCK

        val includeTransactions = isTrue(request.getParameter(INCLUDE_TRANSACTIONS_PARAMETER))

        return JSONData.block(blockData, includeTransactions, blockchain.height, blockService.getBlockReward(blockData), blockService.getScoopNum(blockData))
    }
}
